compiler/rustc_infer/src/infer/error_reporting/mod.rs

```diff
@@ -1168,14 +1168,21 @@ impl<'a, 'tcx> TypeErrCtxt<'a, 'tcx> {
         let output1 = sig1.output();
         let output2 = sig2.output();
         let (x1, x2) = self.cmp(output1, output2);
-        if !output1.is_unit() {
+        let only_output_diff = !lifetime_diff
```

Comment:
Why not just show it whenever the outputs differ?

Reply:
yeah, seems the result not bad:
https://github.com/rust-lang/rust/pull/127417/commits/f46c4129e0310b4fb6a75dbe28c2a86e00c9be27